torch/csrc/jit/passes/memory_planning/greedy_util.cpp

```diff
@@ -0,0 +1,83 @@
+#include <torch/csrc/jit/passes/memory_planning/greedy_by_size.h>
+#include <torch/csrc/jit/passes/memory_planning/greedy_util.h>
+#include <torch/csrc/jit/runtime/static/memory_planner.h>
+
+namespace torch {
+namespace jit {
+
+size_t findOffsetWithSmallestGap(
+    UniqueLiveRange unalloced_ulvr,
+    size_t size,
+    std::vector<MemAllocation> ordered_allocations) {
+  size_t best_gap = std::numeric_limits<size_t>::max();
+  c10::optional<size_t> best_offset = c10::nullopt;
+  size_t prev_offset = 0;
+
+  for (const auto& alloc : ordered_allocations) {
+    if (!overlapLiveRange(alloc.ulvr, unalloced_ulvr)) {
```

Comment:
I'm a bit confused here. It looks like we are skipping live ranges that don't overlap with the unallocated live range? 
I was thinking we should be specifically looking for  the ranges that do *not* overlap otherwise we can't reuse that storage?

Reply:
@Krovatkin this is looking for live ranges that overlap in time and then finding the smallest gap *between* these "live" live ranges. this should really be done using something like an interval tree (and an overlap lookup). I'm looking into this but one issue is interval trees don't deal well with duplicated endpoints. In the single-threaded case that's not an issue but in the concurrency case it might pop up.